nc_time_axis/__init__.py

```diff
@@ -67,14 +70,43 @@ def pick_format(self, resolution):
         return _RESOLUTION_TO_FORMAT[resolution]
 
     def __call__(self, x, pos=0):
-        # Note that self.locator.tick_values must be called
-        # before calling this method; otherwise the locator's
-        # resolution attribute will not be defined.
         format_string = self.pick_format(self.locator.resolution)
         dt = cftime.num2date(x, self.time_units, calendar=self.calendar)
         return dt.strftime(format_string)
 
 
+class NetCDFTimeDateFormatter(AutoCFTimeFormatter):
+    def __init__(self, *args, **kwargs):
+        warnings.warn(
+            "NetCDFTimeDateFormatter will be named AutoCFTimeFormatter "
+            "in nc_time_axis version 1.5",
+            FutureWarning,
+        )
+        super(NetCDFTimeDateFormatter, self).__init__(*args, **kwargs)
+
+
+class CFTimeFormatter(mticker.Formatter):
+    """
+    A formatter for explicitly setting the format of a cftime.datetime axis.
+
+    Parameters
+    ----------
+    format : str
+        Format string that can be passed to cftime.datetime.strftime, e.g. "%Y-%m-%d"
+    calendar : str
+        Calendar type of the datetime axis
+    """
+
+    def __init__(self, format, calendar):
+        self.format = format
+        self.time_units = "days since 2000-01-01"
```

Comment:
As far as I can tell these are *always* the time encoding units used for the tick values.  Since this is critical for this to work, we might consider making these a module-level constant.

Reply:
@aulemahal I'm curious if this is your impression from reading the code as well.  Right now this formatter will only work properly if the tick values can always be converted back to dates using units of `"days since 2000-01-01"`; otherwise users would need to pass their `time_units` to the `CFTimeFormatter` constructor, which feels kind of awkward.

While it is *possible* to pass different time units into the `NetCDFTimeDateLocator` class, in practice I think only the default units are ever used (in fact [this line in `NetCDFTimeConverter` in particular](https://github.com/SciTools/nc-time-axis/blob/ad9153554c292dc9a1ca34bab2751fbc44e5af50/nc_time_axis/__init__.py#L364-L366) already seems to have that baked in).  Do you think we should make this more rigid by removing the `date_unit` argument from the `NetCDFTimeDateLocator` constructor, or leave things be for now?

Reply:
That is my impression also, having  `date_units` or `time_units` as variable doesn't do much, as "days since 2000-01-01" is always used. 
However, I don't know how important this is, but controlling the units can be interesting for specific cases where microsecond resolution is needed  and the dates are far from 2000-01-01. See:

```python3
ind = cftime.datetime(1600, 1, 1, 0, 0, 0, 1, calendar='360_day')
outd = cftime.num2date(cftime.date2num(ind, 'days since 2000-01-01', '360_day'), 'days since 2000-01-01', '360_day')
(outd - ind).total_seconds()
>>> -1e-06
```
So the round trip date -> num -> date was made with an error of 1 ms. However. this is the largest error I was able to get, I get overflow errors before it goes above 1 ms. 

Those this justify leaving it as a parameter? I think not.

Reply:
Indeed, we could consider switching to something like `"microseconds since 1970-01-01"` to be more analogous to NumPy microsecond-precision datetime conventions.  This way we could have exact roundtripping of dates.

For now I kept the default units as `"days since 2000-01-01"` and added deprecation warnings where time units arguments are used.  Perhaps once we go through with deprecating the time units arguments we can revisit this, since at that point the time units will be purely a private implementation detail.  

I updated the tests to no longer use those arguments -- in some tests time units other than "2000-01-01" were used.  Interestingly it does change the values that the `MaxNLocator` chooses in some cases.  Again though, the default behavior is still the same, and now the ticks can be set explicitly, so if a user is not happy with them, they have complete control to change them.  

Anyway, have a look and let me know you think things are good to merge!

Reply:
Well, if there is no roundtripping error with `microseconds since ...` and it is defined on the same range of dates (I just checked), I don't see why we would change to that! I feel nobody is never going to need anything else.
But we can wait a future PR to make smoother changes.

Also I think I just realize that this line : https://github.com/SciTools/nc-time-axis/blob/2c8a84a04968fc731c70322b0c707f623eaa2481/nc_time_axis/__init__.py#L165
is broken : it assumes the units are days, which was not ensured since you could change it. Thus, this PR fixes a problem ;).

Reply:
Indeed!  In the existing code they have a check to make sure the units are `"days"` though you can have an arbitrary reference date.

I agree, long term switching to `"microseconds"` would probably be best.  I'll make an issue to track that and merge this PR now.